Add total_bounds of layer to `read_info`

theroggy opened the pull request:
I made `force_feature_count` default to `False`, which is a breaking change for eg. geojson: will return -1 for `read_info()["features"]` now by default because determining number features is not fast for this file type. 
Not sure if this is OK? If OK, I'll add it as a breaking change in the changelog.

closes #274

brendan-ward reviewed (commented):
Thanks @theroggy !

What do you think about exposing a couple items to the capabilities list as well as [suggested in 274](https://github.com/geopandas/pyogrio/issues/274#issuecomment-1687281701):
* `fast_total_bounds`: `OGR_L_TestCapability(ogr_layer, OLCFastGetExtent)`
* `fast_feature_count`: `OGR_L_TestCapability(ogr_layer, OLCFastFeatureCount)`


I'm OK with adding a breaking change for counts for GeoJSON data sources for `read_info`; this keeps us closer in parity to the underlying driver's capabilities.

theroggy commented on pyogrio/_io.pyx:
```diff
@@ -376,6 +378,33 @@ cdef get_feature_count(OGRLayerH ogr_layer):
+
+    Returns
+    -------
+    Optional[Tuple[float, float, float, float]]
```
ok

kylebarron commented on pyogrio/core.py:
```diff
@@ -148,11 +148,25 @@ def read_bounds(
+    ``features`` will be -1 if determining this is expensive. You can force to calculate
+    it anyway using the ``force_featurecount`` parameter.
+
+    ``total_bounds`` will be None if determining this is expensive. You can force to
```
You probably want `if this is **an** expensive`

theroggy commented on pyogrio/tests/test_core.py:
```diff
@@ -268,19 +270,51 @@ def test_read_info_invalid_dataset_kwargs(naturalearth_lowres):
 
 
-def test_read_info_force_feature_count(data_dir):
+@pytest.mark.parametrize(
```
Yes, this wans't a very pretty test...

theroggy commented:
> What do you think about exposing a couple items to the capabilities list as well as [suggested in 274](https://github.com/geopandas/pyogrio/issues/274#issuecomment-1687281701):
> 
> * `fast_total_bounds`: `OGR_L_TestCapability(ogr_layer, OLCFastGetExtent)`
> * `fast_feature_count`: `OGR_L_TestCapability(ogr_layer, OLCFastFeatureCount)`

Oops, forgot that one... Added them. Apparently the capabilities there had value 0 or 1, which isn't ideal I think? So I changed them to boolean.

> I'm OK with adding a breaking change for counts for GeoJSON data sources for `read_info`; this keeps us closer in parity to the underlying driver's capabilities.

Done. Also added that the capabilities are now booleans.

brendan-ward commented on pyogrio/_io.pyx:
```diff
@@ -376,6 +378,33 @@ cdef get_feature_count(OGRLayerH ogr_layer):
+        The total bounds of the layer, or None if they could not be determined.
+    """
+
+    cdef OGREnvelope ogr_envelope # = NULL
```
```suggestion
    cdef OGREnvelope ogr_envelope = NULL
```

brendan-ward commented on pyogrio/_io.pyx:
```diff
@@ -376,6 +378,33 @@ cdef get_feature_count(OGRLayerH ogr_layer):
+    cdef OGREnvelope ogr_envelope # = NULL
+    try:
+        exc_wrap_ogrerr(OGR_L_GetExtent(ogr_layer, &ogr_envelope, force))
+        bounds = (
```
I imagine there could be a slight though unlikely chance that `ogr_envelope` remains `NULL` without an error being raised, so I'd suggest a check for that before extracting bounds:

```
if ogr_envelope == NULL:
    bounds = None
else:
    bounds = ...

```

theroggy replied:
ok

jorisvandenbossche reviewed (approved):
Looks good to me!
